lib/FileUploadDataSource.ts

```diff
@@ -196,34 +202,45 @@ export default class FileUploadDataSource extends RemoteGraphQLDataSource {
     form.append('map', JSON.stringify(fileMap));
     await this.addDataHandler(form, resolvedFiles);
 
-    const headers = (request.http && request.http.headers) || new Headers();
+    // This must happen before constructing the request headers
+    // otherwise any custom headers set in willSendRequest are ignored
+    if (this.willSendRequest) {
+      await this.willSendRequest(args);
+    }
```

Comment:
test covering this here https://github.com/profusion/apollo-federation-file-upload/pull/67